### PR TITLE
Automatically modify the package manifest to include msbuild for unity

### PR DESCRIPTION
**Pull request by davidkline-ms**
MRTK is taking a dependency on the MSBuild for Unity UPM package to help acquire and manage dependencies.

This change is part of the work to complete #6244 and #6531. It will also enable the unitypackage distribution to acquire the dependency described in #6631 

Testing performed included:
* Updating a manifest with no scoped registries
* Updating a manifest to append to an existing collection of scoped registries.

### CODE REVIEW

File: Assets/MixedRealityToolkit/Utilities/Editor/EditorProjectUtilities.cs [hunk 1]
```diff
@@ -18,6 +18,7 @@ public static class EditorProjectUtilities
         static EditorProjectUtilities()
         {
             CheckMinimumEditorVersion();
+            PackageManifestUpdater.EnsureMSBuildForUnity();
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/Utilities/Editor/EditorProjectUtilities.cs]: Should we make this part of the configurator instead?

File: Assets/MixedRealityToolkit/Utilities/Editor/PackageManifest/PackageManifestUpdater.cs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+                using (StreamReader reader = new StreamReader(manifestStream))
+                {
+                    // Read the manifest file a line at a time.
+                    while (!reader.EndOfStream)
+                    {
+                        string line = reader.ReadLine();
+                        manifestFileLines.Add(line);
+                    }
+
+                    // Go back to the start of the file.q`
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/Utilities/Editor/PackageManifest/PackageManifestUpdater.cs]:
```suggestion
                    // Go back to the start of the file.
```

File: Assets/MixedRealityToolkit/Utilities/Editor/PackageManifest/PackageManifestUpdater.cs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+            // Ensure that pre-exising scoped registries are retained.
+            List<ScopedRegistry> scopedRegistries = new List<ScopedRegistry>();
+            if ((manifest.scopedRegistries != null) && (manifest.scopedRegistries.Length > 0))
+            {
+                scopedRegistries.AddRange(manifest.scopedRegistries);
+            }
+
+            // Attempt to find an entry in the scoped regstries collection for the MSBuild for Unity url
+            string msBuildRegistryUrl = "https://pkgs.dev.azure.com/UnityDeveloperTools/MSBuildForUnity/_packaging/UnityDeveloperTools/npm/registry";
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/Utilities/Editor/PackageManifest/PackageManifestUpdater.cs]: make this a const or readonly?

File: Assets/MixedRealityToolkit/Utilities/Editor/PackageManifest/PackageManifestUpdater.cs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+            foreach (ScopedRegistry registry in scopedRegistries)
+            {
+                if (registry.url == msBuildRegistryUrl)
+                {
+                    needToAddRegistry = false;
+                }
+            }
+
+            // If no entry was found, add one.
+            string msBuildRegistryName = "MS Build for Unity";
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/Utilities/Editor/PackageManifest/PackageManifestUpdater.cs]: make this a const or readonly?

File: Packages/manifest.json [hunk 5]
```diff
@@ -1,5 +1,15 @@
...
+        {
+            "name": "MS Build for Unity",
+            "url": "https://pkgs.dev.azure.com/UnityDeveloperTools/MSBuildForUnity/_packaging/UnityDeveloperTools/npm/registry",
+            "scopes": [
+                "com.microsoft"
+            ]
+        }
+    ],
   "dependencies": {
+"com.microsoft.msbuildforunity": "0.8.1",
```
Comment (keveleigh) [inline Packages/manifest.json]:
```suggestion
    "com.microsoft.msbuildforunity": "0.8.1",
```

Comment (keveleigh) [reply Packages/manifest.json]: Is this how it gets written in? Any way to correct the indentation?

File: Assets/MixedRealityToolkit/Utilities/Editor/EditorProjectUtilities.cs (hunk 1 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/Utilities/Editor/EditorProjectUtilities.cs]: not sure. what is the consensus from the reviewers?

File: Assets/MixedRealityToolkit/Utilities/Editor/PackageManifest/PackageManifestUpdater.cs (hunk 4 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/Utilities/Editor/PackageManifest/PackageManifestUpdater.cs]: not a bad idea

File: Assets/MixedRealityToolkit/Utilities/Editor/PackageManifest/PackageManifestUpdater.cs (hunk 3 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/Utilities/Editor/PackageManifest/PackageManifestUpdater.cs]: not a bad idea

File: Assets/MixedRealityToolkit/Utilities/Editor/EditorProjectUtilities.cs (hunk 1 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit/Utilities/Editor/EditorProjectUtilities.cs]: My thought is making this opt-in. We can gate everything around the presence of the adapter, with remoting / additional feature support (like autoplanar) gated behind an opt-in.

File: Packages/manifest.json (hunk 5 above)
Comment (davidkline-ms) [reply Packages/manifest.json]: we can explicity write the indentation Unity's file currently supports. there is no functional difference with or without the indentation

File: Assets/MixedRealityToolkit/Utilities/Editor/PackageManifest/PackageManifestUpdater.cs [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+            }
+
+            PackageManifest manifest = null;
+
+            // Read the package manifest into a list of strings (for easy finding of entries)
+            // and then deserialize.
+            List<string> manifestFileLines = new List<string>();
+            using (FileStream manifestStream = new FileStream(manifestPath, FileMode.Open, FileAccess.Read))
+            {
+                using (StreamReader reader = new StreamReader(manifestStream))
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/Utilities/Editor/PackageManifest/PackageManifestUpdater.cs]: Do we need the reader for this or could we use [`File.ReadAllLines`](https://docs.microsoft.com/en-us/dotnet/api/system.io.file.readalllines?view=netframework-4.8)?

File: Assets/MixedRealityToolkit/Utilities/Editor/PackageManifest/PackageManifestUpdater.cs [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+                    {
+                        string line = reader.ReadLine();
+                        manifestFileLines.Add(line);
+                    }
+
+                    // Go back to the start of the file.
+                    manifestStream.Seek(0, 0);
+
+                    // Deserialize the scoped registries portion of the package manifest.
+                    manifest = JsonUtility.FromJson<PackageManifest>(reader.ReadToEnd());
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/Utilities/Editor/PackageManifest/PackageManifestUpdater.cs]: Similar question but [`File.ReadAllText`](https://docs.microsoft.com/en-us/dotnet/api/system.io.file.readalltext?view=netframework-4.8)

File: Assets/MixedRealityToolkit/Utilities/Editor/PackageManifest/PackageManifestUpdater.cs (hunk 6 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/Utilities/Editor/PackageManifest/PackageManifestUpdater.cs]: the reader lets us read each line for later writing one at a time.

File: Assets/MixedRealityToolkit/Utilities/Editor/PackageManifest/PackageManifestUpdater.cs (hunk 7 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/Utilities/Editor/PackageManifest/PackageManifestUpdater.cs]: please see above comment

File: Assets/MixedRealityToolkit/Utilities/Editor/EditorProjectUtilities.cs (hunk 1 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/Utilities/Editor/EditorProjectUtilities.cs]: tracking this with #6723

**Pull request merged**